extensions/legit/porcelain.lisp

```diff
@@ -109,554 +79,76 @@ Mercurial:
-  "Return values:
-  - untracked files (todo)
-  - list of ADDED files
-  - modified files"
-  (loop for line in (str:lines (fossil-porcelain))
-        for parts = (str:words line)
-        for status = (first parts)
-        for file = (second parts)
-        if (equal "ADDED" status)
-          collect (list :file file :type :added) into modified-staged-files
-        if (equal "EDITED" status)
-          collect (list :file file :type :modified) into modified-staged-files
-        if (equal "DELETED" status)
-          collect (list :file file :type :deleted) into modified-staged-files
-        if (equal "UNKNOWN" status)
-          collect file into untracked-files
-        finally (return (values untracked-files
-                                nil
-                                modified-staged-files))))
-
-(defun components ()
-  (case *vcs*
-    (:git (git-components))
-    (:fossil (fossil-components))
-    (:hg (hg-components))
-    (t (porcelain-error "VCS not supported: ~a" *vcs*))))
-
-
 ;;;
 ;;; diff
 ;;;
-(defun git-file-diff (file &key cached)
-  ;; --cached is a synonym for --staged.
-  ;; So it is set only for staged files. From git-components: the 3rd value, modified and staged files.
-  (run-git
-   (concatenate 'list
-                *file-diff-args*
-                (list (format nil "-U~D" *diff-context-lines*))
-                (if cached '("--cached"))
-                (list file))))
-
-(defun hg-file-diff (file &key cached)
-  "Show the diff of staged files (and only them)."
-  (when cached
-    (run-hg (list "diff"
-                  (format nil "-U~D" *diff-context-lines*)
-                  file))
-    ;; files not staged can't be diffed.
-    ;; We could read and display their full content anyways?
-    ))
-
-(defun fossil-file-diff (file &key cached)
-  (declare (ignorable cached))
-  (run-fossil (list "diff" file)))
-
-(defun file-diff (file &key cached)
-  (case *vcs*
-    (:fossil
-     (fossil-file-diff file))
-    (:hg
-     (hg-file-diff file :cached cached))
-    (t
-     (git-file-diff file :cached cached))))
-
-
+(defgeneric file-diff (vcs file &key cached)
+  (:documentation "TODO Document: presumably, returns the string form of the diff"))
+
 ;;;
 ;;; Show commits.
 ;;;
-(defun git-show-commit-diff (ref &key ignore-all-space)
-  (let ((options '()))
-    (when ignore-all-space
-      (cl:push "-w" options))
-    (run-git `("show" ,@options ,ref))))
-
-(defun hg-show-commit-diff (ref)
-  (run-hg (list "log" "-r" ref "-p")))
-
-(defun show-commit-diff (ref &key ignore-all-space)
-  (case *vcs*
-    (:fossil nil)
-    (:hg (hg-show-commit-diff ref))
-    (t (git-show-commit-diff ref :ignore-all-space ignore-all-space))))
+(defgeneric show-commit-diff (vcs ref &key ignore-all-space))
 
 ;; commit
-(defun git-commit (message)
-  (run-git (list "commit" "-m" message)))
+(defgeneric commit (vcs message)
+  (:documentation "Performs a commit operation: `message` must be a string."))
```

Comment:
I think a default method that throws an "unsupported/unimplemented" porcelain error for all defgenerics is a good idea.

Reply:
Done. With the amount of repetition in the code though, maybe a case for a macro :sweat_smile: